File: backend/app/schemas/diagnosis.py
"""Pydantic schemas for diagnosis API. Derived from diagnosis.example.py."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field, model_validator


SystemCode = Literal[
    "ENGINE_OIL",
    "ENGINE_COOLING",
    "LANDING_GEAR",
    "MAIN_ROTOR",
    "HYDRAULIC",
    "ELECTRICAL",
    "DRIVE_SYSTEM",
    "FUEL",
    "UNKNOWN",
]

RiskLevel = Literal["LOW", "MEDIUM", "HIGH"]

ALLOWED_VIEW_TARGETS = {
    "AIRCRAFT_OVERVIEW",
    "ENGINE_OIL_SYSTEM",
    "ENGINE_ACCESS_PANEL",
    "ENGINE_INTERNAL_VIEW",
    "ENGINE_OIL_INTERNAL",
    "ENGINE_OIL_FILTER",
    "ENGINE_OIL_PUMP",
    "ENGINE_PRESSURE_SENSOR",
    "HYDRAULIC_SYSTEM",
    "ELECTRICAL_SYSTEM",
    "DRIVE_SYSTEM_VIEW",
    "FUEL_SYSTEM_VIEW",
}


class QueryRequest(BaseModel):
    aircraft_id: str = Field(examples=["DEMO-KUH-01"])
    query: str = Field(min_length=2, max_length=500)
    demo_mode: bool = False


class SourceRef(BaseModel):
    """Public TM citation for DEMO / Public Technical Manual Reference."""

    manual_id: str
    document_id: str
    title: str
    """Compat: same as pdf_page (1-based local PDF index)."""
    page: int | None = None
    """1-based page index in the local PDF file (not the printed page)."""
    pdf_page: int | None = None
    """Task / procedure id from the manual (e.g. 8-3.3)."""
    task: str | None = None
    """Legacy alias for task."""
    paragraph: str | None = None
    """Printed page if verified in PDF footer; never estimated."""
    printed_page: int | None = None

    @model_validator(mode="after")
    def sync_page_fields(self) -> SourceRef:
        if self.pdf_page is None and self.page is not None:
            self.pdf_page = self.page
        if self.page is None and self.pdf_page is not None:
            self.page = self.pdf_page
        if self.task is None and self.paragraph is not None:
            self.task = self.paragraph
        if self.paragraph is None and self.task is not None:
            self.paragraph = self.task
        return self


class DiagnosisResult(BaseModel):
    system_code: SystemCode
    symptom_code: str
    risk_level: RiskLevel
    suspected_components: list[str]
    answer: str
    manual_ids: list[str]
    recommended_steps: list[str]
    view_target_id: str
    confidence: float = Field(ge=0.0, le=1.0)
    is_demo: bool = False
    sources: list[SourceRef] = Field(default_factory=list)
    fault_code: str | None = None
    area_id: str | None = None
    target_mesh: str | None = None
    cover_mesh: str | None = None
    inspection_point: str | None = None
    td_grade: str | None = None

    def assert_view_target_allowed(self) -> None:
        if self.view_target_id not in ALLOWED_VIEW_TARGETS:
            raise ValueError(f"view_target_id not allowed: {self.view_target_id}")


class MaintenanceResultRequest(BaseModel):
    aircraft_id: str
    actions: str
    parts_used: str | None = None
    outcome: str
    query_id: str | None = None


class DemoResetRequest(BaseModel):
    aircraft_id: str | None = "DEMO-KUH-01"

File: frontend/src/types/diagnosis.ts
export type ViewLevel = "AIRCRAFT" | "SYSTEM" | "COMPONENT";
/** Digital-twin drill-down depth (Aircraft → System → Assembly → Component). */
export type InspectionLevel =
  | "EXTERIOR"
  | "SYSTEM"
  | "ASSEMBLY"
  | "COMPONENT";
export type InspectionAssembly =
  | "ENGINE_LEFT"
  | "ENGINE_RIGHT"
  | "GENERATOR_ASSEMBLY"
  | "HYDRAULIC_ASSEMBLY"
  | null;

/** 3D Digital Twin active maintenance system (ELECTRICAL diagnosis → GENERATOR). */
export type ActiveMaintenanceSystem =
  | "ENGINE_OIL"
  | "HYDRAULIC"
  | "GENERATOR";
export type RiskLevel = "LOW" | "MEDIUM" | "HIGH";
export type SystemCode =
  | "ENGINE_OIL"
  | "ENGINE_COOLING"
  | "LANDING_GEAR"
  | "MAIN_ROTOR"
  | "HYDRAULIC"
  | "ELECTRICAL"
  | "DRIVE_SYSTEM"
  | "FUEL"
  | "UNKNOWN";
export type ComponentCode =
  | "OIL_FILTER"
  | "OIL_PUMP"
  | "PRESSURE_SENSOR"
  | "ACCESS_PANEL"
  | "UNKNOWN";

export type BottomTab = "manual" | "failure" | "guide" | "history";

export interface SourceRef {
  manual_id: string;
  document_id: string;
  title: string;
  /** Compat alias of pdf_page */
  page?: number;
  /** 1-based local PDF page index (not printed page) */
  pdf_page?: number | null;
  /** Task id e.g. 8-3.3 */
  task?: string | null;
  /** Legacy alias of task */
  paragraph?: string | null;
  /** Printed page only if verified; else null */
  printed_page?: number | null;
}

export interface DiagnosisResult {
  system_code: SystemCode;
  symptom_code: string;
  risk_level: RiskLevel;
  suspected_components: string[];
  answer: string;
  manual_ids: string[];
  recommended_steps: string[];
  view_target_id: string;
  confidence: number;
  is_demo: boolean;
  sources?: SourceRef[];
  /** Optional structured 3D linkage (AREA mapping). */
  fault_code?: string;
  area_id?: string;
  target_mesh?: string;
  cover_mesh?: string;
  inspection_point?: string;
  /** COMPONENT | SYSTEM | AREA | PENDING */
  td_grade?: string;
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant" | "system";
  text: string;
  time: string;
}

export interface ViewTargetConfig {
  level: ViewLevel;
  label: string;
  cameraPosition: [number, number, number];
  cameraTarget: [number, number, number];
  highlightObjects: string[];
  hideObjects: string[];
  transparentObjects: string[];
  openPanels: string[];
  duration: number;
}

export interface ManualChunk {
  id: string;
  title: string;
  chapter: string;
  section: string;
  page: number;
  content: string;
  system_code: string;
  is_dummy: boolean;
  score?: number;
  document_id?: string;
  paragraph?: string | null;
  task?: string | null;
  source_manual?: string | null;
  original_text?: string | null;
  symptom_ko?: string | null;
  symptom_en?: string | null;
  possible_causes?: string[];
  recommended_steps?: string[];
  component_codes?: string[];
  is_demo_public_tm?: boolean;
  td_mapping?: string | null;
  td_grade?: string | null;
  /** Filename under public/manuals/ (e.g. TM_55-1520-240-T-2.pdf) */
  pdf_file?: string | null;
  /** 1-based PDF Viewer page index (not printed page) */
  pdf_page?: number | null;
  /** Printed page only if verified against the paper TM */
  printed_page?: number | string | null;
  /** Path under public/ for Hybrid page image (e.g. manual-pages/T2_p114.webp) */
  page_image?: string | null;
}

export interface FailureCase {
  id: string;
  symptom: string;
  cause: string;
  actions: string;
  result: string;
  similarity: number;
  is_dummy: boolean;
}

export interface GuideStep {
  n: number;
  title: string;
  detail: string;
  viewTargetId: string;
}

export interface MaintenanceResultItem {
  id: string;
  aircraft_id: string;
  actions: string;
  parts_used?: string;
  outcome: string;
  created_at: string;
}

export interface AircraftModelProps {
  selectedObjectIds: string[];
  hiddenObjectIds: string[];
  transparentObjectIds: string[];
  openedPanelIds: string[];
  onObjectClick: (objectId: string) => void;
}
